src/main/java/org/lanternpowered/server/resource/FileSystemPack.java

```diff
@@ -0,0 +1,223 @@
+                    final String namespace = namespacePath.getFileName().toString();
+                    // Check if the namespace is valid, don't want to break things
+                    if (LanternResourcePath.NAMESPACE_PATTERN.matcher(namespace).matches()) {
+                        Lantern.getLogger().warn("Found a invalid namespace {} within the {} pack.", namespace, getName().toPlain());
+                    } else {
+                        final LanternResourcePath newPath = LanternResourcePath.uncheckedOf(namespace, path.getPath());
+                        final Stream<ResourcePath> extraStream = walkResourcePaths(
+                                newPath, namespacePath.resolve(path.getPath()), maxDepth);
+                        if (extraStream != null) {
+                            stream = stream == null ? extraStream : Stream.concat(stream, extraStream);
+                        }
+                    }
+                }
+            } else {
+                stream = walkResourcePaths(path1, this.dataRoot.resolve(path.getNamespace()).resolve(path.getPath()), maxDepth);
+            }
+        } catch (IOException e) {
+            Lantern.getLogger().error("Failed to walk resource path: {}", path, e);
+        }
+        return stream == null ? Stream.empty() : stream;
+    }
+
+    @Nullable
+    private Stream<ResourcePath> walkResourcePaths(LanternResourcePath resourcePath, Path start, int maxDepth) {
+        if (!Files.exists(start)) {
+            return null;
+        }
+        try {
+            return Files.walk(start, maxDepth).map(path -> {
+                final Path relPath = path.relativize(start);
+                // Backslashes are not supported so replace them
+                final String path1 = relPath.toString().replace('\\', '/');
+                // Check if it's a invalid path for a resource and ignore .mcmeta files
+                if (path1.endsWith(".mcmeta") ||
+                        !LanternResourcePath.PATH_PATTERN.matcher(path1).matches()) {
+                    return null;
+                }
+                // Create the new path
+                return LanternResourcePath.uncheckedOf(resourcePath.getNamespace(),
+                        resourcePath.getPath() + '/' + path1);
+            });
+        } catch (IOException e) {
+            Lantern.getLogger().error("Failed to walk resource path: {}", resourcePath, e);
+        }
+        return null;
+    }
+
+    @Override
+    public Stream<LanternResource> walkResources(ResourcePath path, int maxDepth) {
+        checkNotNull(path, "path");
+        final LanternResourcePath path1 = (LanternResourcePath) path;
+        if (path1.isFile() || maxDepth <= 0) {
+            return Stream.empty();
+        }
+        Stream<LanternResource> stream = null;
+        try {
+            // If a wildcard is used, there is some special handling required
+            if (path1.getNamespace().isEmpty()) {
```

Comment:
This will never be empty. If the namespace is omitted from vanilla, `minecraft` is used in its place.

e.g. `:.` will become `minecraft:.`

Reply:
We could modify that behavior? Or use a wildcard like `*` instead.